File: Source/Core/Configurations/Language.vb
Imports System.IO
Imports System.Xml.Serialization
Imports Mirage.Basic.Engine

Public Class LanguageDef

    Public Load As New LoadDef

    Public Class LoadDef

        Public Loading As String = "Loading..."
        Public Graphics As String = "Initialising Graphics.."
        Public Network As String = "Initialising Network..."
        Public Starting As String = "Starting Game..."

    End Class

    Public MainMenu As New MainMenuDef
    Public Class MainMenuDef

        ' Main Panel
        Public ServerStatus As String = "Server Status:"
        Public ServerOnline As String = "Online"
        Public ServerReconnect As String = "Reconnecting..."
        Public ServerOffline As String = "Offline"
        Public ButtonPlay As String = "Play"
        Public ButtonRegister As String = "Register"
        Public ButtonCredits As String = "Credits"
        Public ButtonExit As String = "Exit"
        Public NewsHeader As String = "Latest News"
        Public News As String = "Welcome To the MirageBasic Client.
                                 This Is a free open source VB.Net game engine!
                                 For help Or support please visit our site at
                                 http://miragebasic.net."

        ' Login Panel
        Public Login As String = "Login"
        Public LoginName As String = "Name : "
        Public LoginPass As String = "Password : "
        Public LoginCheckBox As String = "Save Password?"
        Public LoginButton As String = "Submit"

        ' New Character Panel
        Public NewCharacter As String = "Create Character"
        Public NewCharacterName As String = "Name : "
        Public NewCharacterClass As String = "Class : "
        Public NewCharacterGender As String = "Gender : "
        Public NewCharacterMale As String = "Male"
        Public NewCharacterFemale As String = "Female"
        Public NewCharacterSprite As String = "Sprite"
        Public NewCharacterButton As String = "Submit"

        ' Character Select
        Public UseCharacter As String = "Character Selection"
        Public UseCharacterNew As String = "New Character"
        Public UseCharacterUse As String = "Use Character"
        Public UseCharacterDel As String = "Delete Character"

        ' Register
        Public Register As String = "Registration"
        Public RegisterName As String = "Username : "
        Public RegisterPass1 As String = "Password : "
        Public RegisterPass2 As String = "Retype Password : "

        ' Credits
        Public Credits As String = "Credits"

        ' Misc
        Public StringLegal As String = "You cannot use high ASCII characters In your name, please re-enter."
        Public SendLogin As String = "Connected, sending login information..."
        Public SendNewCharacter As String = "Connected, sending character data..."
        Public SendRegister As String = "Connected, sending registration information..."
        Public ConnectToServer As String = "Connecting to Server...( {0} )"
    End Class

    Public Game As New GameDef
    Public Class GameDef
        Public MapName As String = ""
        Public Time As String = "Time: "
        Public Fps As String = "Fps: "
        Public Lps As String = "Lps: "

        Public Ping As String = "Ping: "
        Public PingSync As String = "Sync"
        Public PingLocal As String = "Local"

        Public MapReceive As String = "Recieving map..."
        Public DataReceive As String = "Receiving game data..."

        Public MapCurMap As String = "Map # {0}"
        Public MapCurLoc As String = "Loc() x: {0} y: {1}"
        Public MapLoc As String = "Cur Loc x: {0} y: {1}"

        Public Fullscreen As String = "Please restart the client for the changes to take effect."
    End Class

    Public Chat As New ChatDef
    Public Class ChatDef

        ' Universal
        Public Emote As String = "Usage : /emote [1-11]"
        Public Info As String = "Usage : /info [player]"
        Public Party As String = "Usage : /party [player]"
        Public PlayerMsg As String = "Usage : ![player] [message]"
        Public HouseInvite As String = "Usage : /houseinvite [player]"
        Public InvalidCmd As String = "Not a valid command!"
        Public Help1 As String = "Social Commands : "
        Public Help2 As String = "'[message] = Global Message"
        Public Help3 As String = "-[message] = Party Message"
        Public Help4 As String = "![player] [message] = Player Message"
        Public Help5 As String = "Available Commands: /help, /info, 
                                  /fps, /lps, /stats, /trade, 
                                  /party, /join, /leave"
                              

        ' Admin-Only
        Public AccessAlert As String = "You need a higher access to do this!"
        Public AdminGblMsg As String = "''msghere = Global Admin Message"
        Public AdminPvtMsg As String = "= msghere = Private Admin Message"
        Public Admin1 As String = "Social Commands:"
        Public Admin2 As String = "Available Commands: /admin, /who, /access, /loc, 
                                   /warpmeto, /warptome, /warpto, 
                                   /sprite, /mapreport, /kick, 
                                   /ban, /respawn, /welcome, /questreset"

        Public Welcome As String = "Usage : /welcome [message]"
        Public Access As String = "Usage : /access [player] [access]"
        Public Sprite As String = "Usage : /sprite [index]"
        Public Kick As String = "Usage : /kick [player]"
        Public Ban As String = "Usage : /ban [player]"

        Public WarpMeTo As String = "Usage : /warpmeto [player]"
        Public WarpToMe As String = "Usage : /warptome [player]"
        Public WarpTo As String = "Usage : /warpto [map index]"

        Public ResetQuest As String = "Usage : /questreset [index]"

        Public InvalidMap As String = "Invalid map index."
        Public InvalidQuest As String = "Invalid quest index."

    End Class

    Public ItemDescription As New ItemDescriptionDef

    Public Class ItemDescriptionDef

        Public NotAvailable As String = "Not Available"
        Public None As String = "None"
        Public Seconds As String = "Seconds"

        Public Currency As String = "Currency"
        Public CommonEvent As String = "Event"
        Public Furniture As String = "Furniture"
        Public Potion As String = "Potion"
        Public Skill As String = "Skill"

        Public Weapon As String = "Weapon"
        Public Armor As String = "Armor"
        Public Helmet As String = "Helmet"
        Public Shield As String = "Shield"
        Public Shoes As String = "Shoes"
        Public Gloves As String = "Gloves"

        Public Amount As String = "Amount : "
        Public Restore As String = "Restore Amount : "
        Public Damage As String = "Damage : "
        Public Defense As String = "Defense : "

    End Class

    Public SkillDescription As New SkillDescriptionDef
    Public Class SkillDescriptionDef

        Public No As String = "No"
        Public None As String = "None"
        Public Warp As String = "Warp"
        Public Tiles As String = "Tiles"
        Public SelfCast As String = "Self-Cast"

        Public Gain As String = "Regen : "
        Public GainHp As String = "Regen HP"
        Public GainMp As String = "Regen MP"
        Public Lose As String = "Syphon : "
        Public LoseHp As String = "Syphon HP"
        Public LoseMp As String = "Syphon MP"

    End Class

    Public Crafting As New CraftingDef
    Public Class CraftingDef

        Public NotEnough As String = "Not enough materials!"
        Public NotSelected As String = "Nothing selected"

    End Class

    Public Trade As New TradeDef
    Public Class TradeDef

        Public Request As String = "{0} is requesting to trade."
        Public Timeout As String = "You took too long to decide. Please try again."

        Public Value As String = "Total Value : {0}g"

        Public StatusOther As String = "Other player confirmed offer."
        Public StatusSelf As String = "You confirmed the offer."

    End Class

    Public Events As New EventDef
    Public Class EventDef

        Public OptContinue = "- Continue -"

    End Class

    Public Quest As New QuestDef
    Public Class QuestDef

        Public Cancel As String = "Cancel Started"
        Public Started As String = "Quest Started"
        Public Completed As String = "Quest Completed"

        ' Quest Types
        Public Slay As String = "Defeat {0}/{1} {2}."
        Public Collect As String = "Collect {0}/{1} {2}."
        Public Talk As String = "Go talk To {0}."
        Public Reach As String = "Go To {0}."
        Public TurnIn As String = "Give {0} the {1} {2}/{3} they requested."
        Public Kill As String = "Defeat {0}/{1} Players In Battle."
        Public Gather As String = "Gather {0}/{1} {2}."
        Public Fetch As String = "Fetch {0} X {1} from {2}."

    End Class

    Public Character As New CharacterDef
    Public Class CharacterDef

        Public PName As String = "Name: "
        Public JobType As String = "Job: "
        Public Level As String = "Lv: "
        Public Exp As String = "Exp: "

        Public StatsLabel As String = "Stats:"
        Public Strength As String = "Strength: "
        Public Endurance As String = "Endurance: "
        Public Vitality As String = "Vitality: "
        Public Intelligence As String = "Intelligence: "
        Public Luck As String = "Luck: "
        Public Spirit As String = "Spirit: "
        Public Points As String = "Points Available: "

        Public SkillLabel As String = "Skills:"

    End Class
End Class

Public Module modLanguage
    Public Language As New LanguageDef

    Public Sub LoadLanguage()
        Dim cf As String = Paths.Config()
        Dim x As New XmlSerializer(GetType(LanguageDef), New XmlRootAttribute("Language"))

        Directory.CreateDirectory(cf)
        cf += "language.xml"

        If Not File.Exists(cf) Then
            File.Create(cf).Dispose()
            Dim writer = New StreamWriter(cf)
            x.Serialize(writer, Language)
            writer.Close
        End If

        Dim reader = New StreamReader(cf)
        Language = x.Deserialize(reader)
        reader.Close
    End Sub
End Module

File: Source/Core/Configurations/Language.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Xml.Serialization;

namespace Core
{

    public class Locales
    {
        public string GetValueByKey(List<KeyValuePair<string, string>> list, string key)
        {
            var pair = list.FirstOrDefault(kvp => kvp.Key == key);
            return pair.Equals(default(KeyValuePair<string, string>)) ? "Key not found" : pair.Value;
        }

        public string GetValueByKey(string listName, string key)
        {
            var property = this.GetType().GetProperty(listName);
            if (property == null)
                return "List not found";
            var list = (List<KeyValuePair<string, string>>)property.GetValue(this);
            return GetValueByKey(list, key);
        }

        public List<KeyValuePair<string, string>> Load { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("Loading", "Loading..."),
                new KeyValuePair<string, string>("Graphics", "Initializing Graphics.."),
                new KeyValuePair<string, string>("Network", "Initializing Network..."),
                new KeyValuePair<string, string>("Starting", "Starting Game...")
            };

        public List<KeyValuePair<string, string>> MainMenu { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("ServerStatus", "Server Status:"),
                new KeyValuePair<string, string>("ServerOnline", "Online"),
                new KeyValuePair<string, string>("ServerReconnect", "Reconnecting..."),
                new KeyValuePair<string, string>("ServerOffline", "Offline"),
                new KeyValuePair<string, string>("ButtonPlay", "Play"),
                new KeyValuePair<string, string>("ButtonRegister", "Register"),
                new KeyValuePair<string, string>("ButtonCredits", "Credits"),
                new KeyValuePair<string, string>("ButtonExit", "Exit"),
                new KeyValuePair<string, string>("NewsHeader", "Latest News"),
                new KeyValuePair<string, string>("News", @"Welcome To the XtremeWorlds.
                               This is a free open-source C# game engine!
                               For help or support please visit our site at
                               https://xtremeworlds.com/."),
                new KeyValuePair<string, string>("Login", "Login"),
                new KeyValuePair<string, string>("LoginName", "Name: "),
                new KeyValuePair<string, string>("LoginPass", "Password: "),
                new KeyValuePair<string, string>("LoginCheckBox", "Save Password?"),
                new KeyValuePair<string, string>("LoginButton", "Submit"),
                new KeyValuePair<string, string>("NewCharacter", "Create Character"),
                new KeyValuePair<string, string>("NewCharacterName", "Name: "),
                new KeyValuePair<string, string>("NewCharacterClass", "Class: "),
                new KeyValuePair<string, string>("NewCharacterGender", "Gender: "),
                new KeyValuePair<string, string>("NewCharacterMale", "Male"),
                new KeyValuePair<string, string>("NewCharacterFemale", "Female"),
                new KeyValuePair<string, string>("NewCharacterSprite", "Sprite"),
                new KeyValuePair<string, string>("NewCharacterButton", "Submit"),
                new KeyValuePair<string, string>("UseCharacter", "Character Selection"),
                new KeyValuePair<string, string>("UseCharacterNew", "New Character"),
                new KeyValuePair<string, string>("UseCharacterUse", "Use Character"),
                new KeyValuePair<string, string>("UseCharacterDel", "Delete Character"),
                new KeyValuePair<string, string>("Register", "Registration"),
                new KeyValuePair<string, string>("RegisterName", "Username: "),
                new KeyValuePair<string, string>("RegisterPass1", "Password: "),
                new KeyValuePair<string, string>("RegisterPass2", "Retype Password: "),
                new KeyValuePair<string, string>("Credits", "Credits"),
                new KeyValuePair<string, string>("StringLegal", "You cannot use high ASCII characters In your name, please re-enter."),
                new KeyValuePair<string, string>("SendLogin", "Connected, sending login information..."),
                new KeyValuePair<string, string>("SendNewCharacter", "Connected, sending character data..."),
                new KeyValuePair<string, string>("SendRegister", "Connected, sending registration information..."),
                new KeyValuePair<string, string>("ConnectToServer", "Connecting to Server...( {0} )")
            };

        public List<KeyValuePair<string, string>> Game { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("Time", "Time: "),
                new KeyValuePair<string, string>("Fps", "Fps: "),
                new KeyValuePair<string, string>("Lps", "Lps: "),
                new KeyValuePair<string, string>("Ping", "Ping: "),
                new KeyValuePair<string, string>("PingSync", "Sync"),
                new KeyValuePair<string, string>("PingLocal", "Local"),
                new KeyValuePair<string, string>("MapReceive", "Receiving map..."),
                new KeyValuePair<string, string>("DataReceive", "Receiving game data..."),
                new KeyValuePair<string, string>("MapCurMap", "Map # {0}"),
                new KeyValuePair<string, string>("MapCurLoc", "Loc() x: {0} y: {1}"),
                new KeyValuePair<string, string>("MapLoc", "Cur Loc x: {0} y: {1}"),
                new KeyValuePair<string, string>("Fullscreen", "Please restart the client for the changes to take effect."),
                new KeyValuePair<string, string>("InvalidMap", "Invalid map index."),
                new KeyValuePair<string, string>("AccessDenied", "Access Denied!")
            };

        public List<KeyValuePair<string, string>> Chat { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("Emote", "Usage : /emote [1-11]"),
                new KeyValuePair<string, string>("Info", "Usage : /info [player]"),
                new KeyValuePair<string, string>("Party", "Usage : /party [player]"),
                new KeyValuePair<string, string>("Trade", "Usage : /trade [player]"),
                new KeyValuePair<string, string>("PlayerMsg", "Usage : ![player] [message]"),
                new KeyValuePair<string, string>("InvalidCmd", "Not a valid command!"),
                new KeyValuePair<string, string>("Help1", "Social Commands : "),
                new KeyValuePair<string, string>("Help2", "'[message] = Global Message"),
                new KeyValuePair<string, string>("Help3", "-[message] = Party Message"),
                new KeyValuePair<string, string>("Help4", "![player] [message] = Player Message"),
                new KeyValuePair<string, string>("Help5", "@[message] = Admin Message"),
                new KeyValuePair<string, string>("Help6", @"Available Commands: /help, /info, 
                                /fps, /lps, /stats, /trade, 
                                /party, /join, /leave"),
                new KeyValuePair<string, string>("AdminGblMsg", "''msghere = Global Admin Message"),
                new KeyValuePair<string, string>("AdminPvtMsg", "= msghere = Private Admin Message"),
                new KeyValuePair<string, string>("Admin1", "Social Commands:"),
                new KeyValuePair<string, string>("Admin2", @"Available Commands: /admin, /who, /access, /loc, 
                                /warpmeto, /warptome, /warpto, 
                                /sprite, /mapreport, /kick, 
                                /ban, /respawn, /welcome,
                                /editmap, /edititem, /editresource,
                                /editskill, /editpet, /editshop
                                /editprojectile, /editnpc, /editjob
                                /editjob, /acp"),
                new KeyValuePair<string, string>("Welcome", "Usage : /welcome [message]"),
                new KeyValuePair<string, string>("Access", "Usage : /access [player] [access]"),
                new KeyValuePair<string, string>("Sprite", "Usage : /sprite [index]"),
                new KeyValuePair<string, string>("Kick", "Usage : /kick [player]"),
                new KeyValuePair<string, string>("Ban", "Usage : /ban [player]"),
                new KeyValuePair<string, string>("WarpMeTo", "Usage : /warpmeto [player]"),
                new KeyValuePair<string, string>("WarpToMe", "Usage : /warptome [player]"),
                new KeyValuePair<string, string>("WarpTo", "Usage : /warpto [map index]"),
            };

        public List<KeyValuePair<string, string>> ItemDescription { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("NotAvailable", "Not Available"),
                new KeyValuePair<string, string>("None", "None"),
                new KeyValuePair<string, string>("Seconds", "Seconds"),
                new KeyValuePair<string, string>("Currency", "Currency"),
                new KeyValuePair<string, string>("CommonEvent", "Event"),
                new KeyValuePair<string, string>("Furniture", "Furniture"),
                new KeyValuePair<string, string>("Potion", "Potion"),
                new KeyValuePair<string, string>("Skill", "Skill"),
                new KeyValuePair<string, string>("Weapon", "Weapon"),
                new KeyValuePair<string, string>("Armor", "Armor"),
                new KeyValuePair<string, string>("Helmet", "Helmet"),
                new KeyValuePair<string, string>("Shield", "Shield"),
                new KeyValuePair<string, string>("Shoes", "Shoes"),
                new KeyValuePair<string, string>("Gloves", "Gloves"),
                new KeyValuePair<string, string>("Amount", "Amount : "),
                new KeyValuePair<string, string>("Restore", "Restore Amount : "),
                new KeyValuePair<string, string>("Damage", "Damage : "),
                new KeyValuePair<string, string>("Defense", "Defense : ")
            };

        public List<KeyValuePair<string, string>> SkillDescription { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("No", "No"),
                new KeyValuePair<string, string>("None", "None"),
                new KeyValuePair<string, string>("Warp", "Warp"),
                new KeyValuePair<string, string>("Tiles", "Tiles"),
                new KeyValuePair<string, string>("SelfCast", "Self-Cast"),
                new KeyValuePair<string, string>("Gain", "Regen : "),
                new KeyValuePair<string, string>("GainHp", "Regen HP"),
                new KeyValuePair<string, string>("GainMp", "Regen MP"),
                new KeyValuePair<string, string>("Lose", "Syphon : "),
                new KeyValuePair<string, string>("LoseHp", "Syphon HP"),
                new KeyValuePair<string, string>("LoseMp", "Syphon MP")
            };

        public List<KeyValuePair<string, string>> Crafting { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("NotEnough", "Not enough materials!"),
                new KeyValuePair<string, string>("NotSelected", "Nothing selected")
            };

        public List<KeyValuePair<string, string>> Trade { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("Request", "{0} is requesting to trade."),
                new KeyValuePair<string, string>("Timeout", "You took too long to decide. Please try again."),
                new KeyValuePair<string, string>("Value", "Total Value : {0}g"),
                new KeyValuePair<string, string>("StatusOther", "Other player confirmed offer."),
                new KeyValuePair<string, string>("StatusSelf", "You confirmed the offer.")
            };

        public List<KeyValuePair<string, string>> Events { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("OptContinue", "- Continue -")
            };

        public List<KeyValuePair<string, string>> Quest { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("Cancel", "Cancel Started"),
                new KeyValuePair<string, string>("Started", "Quest Started"),
                new KeyValuePair<string, string>("Completed", "Quest Completed"),
                new KeyValuePair<string, string>("Slay", "Defeat {0}/{1} {2}."),
                new KeyValuePair<string, string>("Collect", "Collect {0}/{1} {2}."),
                new KeyValuePair<string, string>("Talk", "Go talk To {0}."),
                new KeyValuePair<string, string>("Reach", "Go To {0}."),
                new KeyValuePair<string, string>("TurnIn", "Give {0} the {1} {2}/{3} they requested."),
                new KeyValuePair<string, string>("Kill", "Defeat {0}/{1} Players In Battle."),
                new KeyValuePair<string, string>("Gather", "Gather {0}/{1} {2}."),
                new KeyValuePair<string, string>("Fetch", "Fetch {0} X {1} from {2}.")
            };

        public List<KeyValuePair<string, string>> Character { get; set; } = new List<KeyValuePair<string, string>>
            {
                new KeyValuePair<string, string>("PName", "Name: "),
                new KeyValuePair<string, string>("JobType", "Job: "),
                new KeyValuePair<string, string>("Level", "Lv: "),
                new KeyValuePair<string, string>("Exp", "Exp: "),
                new KeyValuePair<string, string>("StatsLabel", "Stats:"),
                new KeyValuePair<string, string>("Strength", "Strength: "),
                new KeyValuePair<string, string>("Endurance", "Endurance: "),
                new KeyValuePair<string, string>("Vitality", "Vitality: "),
                new KeyValuePair<string, string>("Intelligence", "Intelligence: "),
                new KeyValuePair<string, string>("Luck", "Luck: "),
                new KeyValuePair<string, string>("Spirit", "Spirit: "),
                new KeyValuePair<string, string>("Points", "Points Available: "),
                new KeyValuePair<string, string>("SkillLabel", "Skills:")
            };
    }

    public class LocalesManager
    {
        public static Locales Language = new Locales();

        public static void Load()
        {
            string cf = Path.Config;
            var x = new XmlSerializer(typeof(Locales), new XmlRootAttribute("Language"));

            Directory.CreateDirectory(cf);
            cf += "Language.xml";

            if (!File.Exists(cf))
            {
                File.Create(cf).Dispose();
                var writer = new StreamWriter(cf);
                x.Serialize(writer, Language);
                writer.Close();
            }

            var reader = new StreamReader(cf);
            Language = (Locales)x.Deserialize(reader);
            reader.Close();
        }
    }
}